compiler/codegen/OMRCodeGenPhase.cpp

```diff
@@ -423,23 +423,63 @@ OMR::CodeGenPhase::performSetupForInstructionSelectionPhase(TR::CodeGenerator *
 
    if (cg->isConcurrentScavengeEnabled())
       {
-      // TODO (GuardedStorage): We need to come up with a better solution than anchoring tree tops of compressedrefs
-      // sequences to enforce certain evaluation order. Perhaps not lowering compressedrefs in the first place when
-      // concurrent scavenge is supported is the correct solution here.
+      // TODO (GuardedStorage): We need to come up with a better solution than anchoring aloadi's
+      // to enforce certain evaluation order
       traceMsg(comp, "GuardedStorage: in performSetupForInstructionSelectionPhase\n");
 
       auto mapAllocator = getTypedAllocator<std::pair<TR::TreeTop*, TR::TreeTop*> >(comp->allocator());
 
       std::map<TR::TreeTop*, TR::TreeTop*, std::less<TR::TreeTop*>, TR::typed_allocator<std::pair<TR::TreeTop*, TR::TreeTop*>, TR::Allocator> >
          currentTreeTopToappendTreeTop(std::less<TR::TreeTop*> (), mapAllocator);
 
+      TR_BitVector *unAnchorableAloadiNodes = comp->getBitVectorPool().get();
+
       for (TR::PreorderNodeIterator iter(comp->getStartTree(), comp); iter != NULL; ++iter)
          {
          TR::Node *node = iter.currentNode();
 
          traceMsg(comp, "GuardedStorage: Examining node = %p\n", node);
 
-         if ((comp->useCompressedPointers() && node->getOpCodeValue() == TR::l2a) || (!comp->useCompressedPointers() && node->getOpCodeValue() == TR::aloadi))
+         if (node->getOpCodeValue() == TR::NULLCHK)
+            {
+            // An aloadi cannot be anchored if there is a Null Check on
+            // its child. There are two situations where this occurs.
+            // The first is when doing an aloadi off some node that is
+            // being NULLCHK'd (see Ex1). The second is when doing an
+            // icalli in which case the aloadi loads the VFT of an
+            // object that must be NULLCHK'd (see Ex2).
+            //
+            // Ex1:
+            //    n1n NULLCHK on n3n
+            //    n2n    aloadi f
+            //    n3n       aload O
+            //
+            // Ex2:
+            //    n1n NULLCHK on n4n
+            //    n2n    icall foo
+            //    n3n       aloadi <vft>
+            //    n4n          aload O
+
+            TR::Node *nodeBeingNullChkd = node->getNullCheckReference();
+            TR::Node *firstChild = node->getFirstChild();
+            TR::Node *grandChild = firstChild->getFirstChild();
+            TR::Node *greatGrandChild = grandChild->getFirstChild();
```

Comment:
My understanding is that the grandchild of a NULLCHK should always be defined (at least by the time we get to codegen - @andrewcraik can correct me if I'm wrong). The greatGrandChild might be NULL though. However I will add checks for both the grandChild and the greatGrandChild.